File: game-ui/src/game/piece/Advisor.js
import Piece from "./Piece";

export default class Advisor extends Piece {
    constructor(shortName, isRed) {
        super(shortName, isRed);
    }

    isValidMove(board, from, to) {
        const xFrom = from.x + "";
        const yFrom = from.y + "";
        const xTo = to.x + "";
        const yTo = to.y + "";

        const pieceAtTo = board[xTo][yTo];

        if (yFrom >= 3 && yFrom <= 5 && yTo >= 3 && yTo <= 5) {
            if (Math.abs(xFrom - xTo) === 1 && Math.abs(yFrom - yTo) === 1) {
                if (this.isRed && xFrom >= 7 && xFrom <= 9 && xTo >= 7 && xTo <= 9) {
                    // Just move
                    if (pieceAtTo == null) return true;
                    // Fight (eat piece).
                    return pieceAtTo.isRed !== this.isRed;
                }
                if (!this.isRed && xFrom >= 0 && xFrom <= 2 && xTo >= 0 && xTo <= 2) {
                    // Just move
                    if (pieceAtTo == null) return true;
                    // Fight (eat piece).
                    return pieceAtTo.isRed !== this.isRed;
                }
            }
        }

        return false;
    }

}

File: ccgame-service/src/main/java/com/doubleat/ccgame/game/piece/Advisor.java
package com.doubleat.ccgame.game.piece;

import com.doubleat.ccgame.game.Board;
import com.doubleat.ccgame.game.Position;

public class Advisor extends Piece {
    public Advisor(String shortName) {
        super(shortName);
    }

    @Override
    public boolean isValidMove(Board board, Position from, Position to) {

        int xFrom = from.getX();
        int yFrom = from.getY();
        int xTo = to.getX();
        int yTo = to.getY();

        Piece pieceAtTo = board.getPieces()[xTo][yTo];

        if (yFrom >= 3 && yFrom <= 5 && yTo >= 3 && yTo <= 5) {
            if (Math.abs(xFrom - xTo) == 1 && Math.abs(yFrom - yTo) == 1) {
                if (this.isRed && xFrom >= 7 && xFrom <= 9 && xTo >= 7 && xTo <= 9) {
                    if (pieceAtTo == null) {
                        return true;
                    }
                    return !pieceAtTo.isRed;
                }
                if (!this.isRed && xFrom >= 0 && xFrom <= 2 && xTo <= 2) {
                    // Just move
                    if (pieceAtTo == null)
                        return true;
                    // Fight (eat piece).
                    return pieceAtTo.isRed;
                }
            }
        }
        return false;
    }

}
